File: src/main/java/com/tms/csp/fm/dnnf/visitor/Traversal.java
package com.tms.csp.fm.dnnf.visitor;

import com.tms.csp.ast.Exp;

public class Traversal {

    private static int nextTraversalId;

    public int traversalId;

    public Traversal() {
        nextTraversalId++;
        this.traversalId = nextTraversalId;
    }

    public boolean visited(Exp n) {
        throw new UnsupportedOperationException();
//        return n.traversalId == this.traversalId;
    }

    public void markAsVisited(Exp n) {
        throw new UnsupportedOperationException();
//        n.traversalId = this.traversalId;
    }

}

File: src/main/java/com/tms/csp/fm/dnnf/visitor/Traversal.kt
package com.tms.csp.fm.dnnf.visitor

import com.tms.csp.ast.Exp

open class Traversal {

    var traversalId: Int = 0

    init {
        nextTraversalId++
        this.traversalId = nextTraversalId
    }

    fun visited(n: Exp): Boolean {
        //        throw new UnsupportedOperationException();
        return n.traversalId == this.traversalId
    }

    fun markAsVisited(n: Exp) {
        //        throw new UnsupportedOperationException();
        n.traversalId = this.traversalId
    }

    companion object {

        private var nextTraversalId: Int = 0
    }

}
